src/main.cpp

```diff
@@ -4049,9 +4051,11 @@ bool ProcessNewBlock(CValidationState& state, const CChainParams& chainparams, c
 
         // Store to disk
         CBlockIndex *pindex = NULL;
-        bool ret = AcceptBlock(*pblock, state, chainparams, &pindex, fRequested, dbp);
+        bool fNewBlock = false;
+        bool ret = AcceptBlock(*pblock, state, chainparams, &pindex, fRequested, dbp, &fNewBlock);
         if (pindex && pfrom) {
             mapBlockSource[pindex->GetBlockHash()] = pfrom->GetId();
+            if (fNewBlock) pfrom->nLastBlockTime = GetTime();
```

Comment:
As far as I understand:
- `ret` could be false due to our local issues (so it's ignored here and used later only to throw an error);
- `pindex != NULL` means that at least header is ok (so we save id of the peer who sent this block to us);
- `fNewBlock == true` means that 
   - we don't have data for that block yet;
   - we either requested this block or otherwise it at least a) wasn't pruned earlier, b) has more work than our tip and c) is close enough to our tip.

So `pindex && fNewBlock` means that there is some valuable chain and we'd better keep this peer to have more info about that chain (even if we think that the chain is invalid).